### DISCUSSION TITLE
Presenting rapidly changing numbers

### DISCUSSION THREAD

**Discussion by markusgermar** [ctreffe/alfred#175] [Alfred Q&A] [answered]
I would like to present the numbers 1, 2, and 3 (or other numerical sequences). The presentation should be very fast (i.e., as an fast animation) and at random. I tried using AutoForwardPages in a Section (using for loop and random.choice, etc....) but since timeout must be >= 1s, the presentation is awkwardly slow.

Is there a way to present the numbers rapidly within one page without reloading? Or is there another way?

Thanks!

**Comment by ctreffe** [comment]
Hey Markus,

I think this should be easy to achieve using JavaScript. You can find a description and a minimal example of how to add JavaScript to an alfred3 page in the [documentation](https://alfredo3.psych.bio.uni-goettingen.de/docs/generated/alfred3.element.misc.JavaScript.html?highlight=javascript). The JavaScript element will allow you to rapidly change the text of an alfred3 text element.

Changing the content of an element using JavaScript is quite simple and if you combine this with a _setInterval_ loop you're done. There are plenty of tutorials on related issues for JavaScript, I think you will have no problems to find suitable code snippets. Also, the linked minimal example in the documentation will give you an Idea of how to access alfred3 elements on a page with JavaScript by using their name.

**Comment by jobrachem** [comment, marked as answer]
You can achieve your desired result by including JavaScript code via the `alfred3.JavaScript` element ([docs](https://alfredo3.psych.bio.uni-goettingen.de/docs/generated/alfred3.element.misc.JavaScript.html#alfred3.element.misc.JavaScript)). The name of an alfred3 element becomes the object id in the HTML code on the page, so you can use this to manipulate the object with JavaScript on a single page. Here's an example experiment:

```python
import alfred3 as al
exp = al.Experiment()

js = """
$(document).ready(function() {
    
    var update_number = function() {
        var random_number = Math.random() // draws a random number
        
        $( '#number_display' ).html(random_number) // updates the display
    };

    var interval = 500; // interval length in milliseconds
    setInterval(update_number, interval);
})
"""

@exp.member
class Demo(al.Page):
    
    def on_exp_access(self):
        self += al.Text("-", name="number_display")
        self += al.JavaScript(code=js)
```

This will create the following experiment:

![demo](https://user-images.githubusercontent.com/37882800/149125160-a5ce44e5-acc2-46fd-bfb9-05875536906d.gif)


### How to control the JavaScript?

Basically, you can ...

1. ... change the code that generates the random number. I used `Math.random()` in this case, which gives a random number between 0 and 1.
2. ... select a different element. I selected my text element with `$( '#number_display' )`. If you have a different name, say `different_name`, fill that in here: `$( '#different_name' )`.
3. ... select a different update interval. I used `var interval = 500;`, which is half a second (500 milliseconds).

```js
$(document).ready(function() {
    
    var update_number = function() {
        var random_number = Math.random() // draws a random number
        
        $( '#number_display' ).html(random_number) // updates the display
    };

    var interval = 500; // interval length in milliseconds
    setInterval(update_number, interval);
})
```

**Note that the sequence of random numbers will not be saved to the experiment data frame, as it exists only on the page.**

**Reply by ctreffe** [reply]
Hey, I was here first 😃
